src/lib/Hob/Ui/Editor/Fancy.hs

```diff
@@ -0,0 +1,1006 @@
+            stroke
+            where
+                (Point cursorLeft cursorTop) = cursorPosition dData
+                cursorBottom = cursorTop + fontAscent opts + fontDescent opts
+
+        getDrawableData = do
+            fancyEditorData <- liftIO $ readMVar fancyEditorDataHolder
+            return (drawingData fancyEditorData, drawingOptions fancyEditorData)
+
+getActiveEditor :: Context -> IO (Maybe DrawingArea)
+getActiveEditor = maybe (return Nothing) getEditorFromNotebookTab <=< getActiveEditorTab
+
+getEditorFromNotebookTab :: Widget -> IO (Maybe DrawingArea)
+getEditorFromNotebookTab currentlyActiveEditor =
+    if currentlyActiveEditor `isA` gTypeScrolledWindow then do
+        let textEditScroller = castToScrolledWindow currentlyActiveEditor
+        textEdit <- binGetChild textEditScroller
+        let areaa = (\area -> if isA area gTypeDrawingArea then Just area else Nothing) =<< textEdit
+        return $ fmap castToDrawingArea areaa
+    else return Nothing
+
+getActiveEditorTab :: Context -> IO (Maybe Widget)
+getActiveEditorTab ctx = do
+    pageNum <- notebookGetCurrentPage tabbed
+    if pageNum < 0 then
+        return Nothing
+    else do
+        tabs <- containerGetChildren tabbed
+        return $ Just $ tabs!!pageNum
+    where tabbed = mainNotebook.uiContext $ ctx
+
+tabTitleForFile :: Maybe FilePath -> String
+tabTitleForFile (Just filePath) = filename' filePath
+    where filename' = encodeString . filename . decodeString
+tabTitleForFile Nothing = "(new file)"
+
+setEditorWidgetFilePath :: WidgetClass a => a -> Maybe FilePath -> IO ()
+setEditorWidgetFilePath editor filePath = do
+    quark <- fileNameQuark
+    objectSetAttribute quark editor filePath
+
+setEditorId :: WidgetClass a => a -> Int -> IO ()
+setEditorId editor identifier = do
+    quark <- editorIdQuark
+    objectSetAttribute quark editor $ Just identifier
+
+getEditorId :: WidgetClass a => a -> IO Int
+getEditorId editor = do
+    quark <- editorIdQuark
+    mRet <- objectGetAttributeUnsafe quark editor
+    case mRet of
+        Just ret -> return ret
+        Nothing -> error "editor has no id assigned"
+
+setEditorModes :: WidgetClass a => a -> [Mode] -> IO ()
+setEditorModes editor modes = do
+    quark <- editorModesQuark
+    objectSetAttribute quark editor $ Just modes
+
+getEditorModes :: WidgetClass a => a -> IO [Mode]
+getEditorModes editor = do
+    quark <- editorModesQuark
+    mRet <- objectGetAttributeUnsafe quark editor
+    case mRet of
+        Just ret -> return ret
+        Nothing -> error "editor has no modes assigned"
+{-
+updateEditorTitle :: WidgetClass a => a -> Editor -> IO ()
+updateEditorTitle editorWidget editor = do
+    Just scrolledW <- widgetGetParent editorWidget
+    Just notebookW <- widgetGetParent scrolledW
+    let notebook = castToNotebook notebookW
+    notebookSetTabLabelText notebook scrolledW =<< tabTitleForEditor editor
+-}
+fileNameQuark :: IO Quark
+fileNameQuark = quarkFromString "fileName"
```

Comment:
achievement unlocked. this file will be fun to split / reorganise 
![image](https://cloud.githubusercontent.com/assets/893660/6320367/389b959a-bad3-11e4-9bf8-a45efc9194cc.png)


the few emerged elements:
- source code parser and block collector.
- drawing data generator
- text operations 
- editor api export
- gtk facing code